File: bai/read-balance.py
#!/usr/bin/env python3
"""Остаток B.AI с chat.b.ai/usage — браузером Camoufox.

Почему не HTTP и не Playwright: страница за Cloudflare, и `curl` даже с живыми куками профиля
получает `403 Cf-Mitigated: challenge`, а Playwright упирается в тот же экран и в headless, и в
видимом окне, и с `channel: 'chrome'` (проверено 16.09). Camoufox этот экран проходит.

CF-виджет обходим НЕ своим кодом: в репозитории уже есть рабочий обход — `solve_turnstile()`
в `anymodel/lib/camoufox_anymodel.py`, с человеческим кликом мышью по iframe (DOM-клик по iframe
Turnstile не видит). Если он не пройдёт сам — окно видимое, и рядом сидит человек: проверку
можно пройти руками, дальше профиль носит `cf_clearance` сам.

Печатает одну строку JSON:
  {"ok": true, "balance": 297604, "raw": "297,604"}
  {"ok": false, "reason": "...", "title": "...", "text": "..."}

Запуск:  python bai/read-balance.py [--headless]
"""
import asyncio
import json
import pathlib
import re
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent / 'anymodel' / 'lib'))

from camoufox.async_api import AsyncCamoufox          # noqa: E402
from camoufox_anymodel import solve_turnstile          # noqa: E402  — переиспользуем обход CF

PROFILE = pathlib.Path(__file__).resolve().parent / 'camoufox-profile'
URL = 'https://chat.b.ai/usage'

# 🪤 Цепляемся не за класс (у них генерированные `acss-1h7d5nm`, меняются от сборки к сборке),
# а за СОСЕДА элемента с текстом ровно «Balance».
READ_JS = """
() => {
  const label = [...document.querySelectorAll('div')]
      .find(n => n.children.length === 0 && n.textContent.trim() === 'Balance');
  const txt = label && label.nextElementSibling && label.nextElementSibling.textContent.trim();
  return txt && /\\d/.test(txt) ? txt : null;
}
"""


async def read_once(page):
    try:
        return await page.evaluate(READ_JS)
    except Exception:
        return None


async def main():
    headless = '--headless' in sys.argv
    async with AsyncCamoufox(
        headless=headless,
        os='windows',
        window=(1280, 900),
        persistent_context=True,
        user_data_dir=str(PROFILE),
        disable_coop=True,
        humanize=True,
        main_world_eval=True,
        i_know_what_im_doing=True,
    ) as browser:
        page = browser.pages[0] if browser.pages else await browser.new_page()
        await page.goto(URL, wait_until='domcontentloaded', timeout=90000)

        raw = None
        for attempt in range(8):
            raw = await read_once(page)
            if raw:
                break
            # Экрана нет, но и числа нет — вероятно, ещё грузится; даём странице время.
            await asyncio.sleep(3)
            raw = await read_once(page)
            if raw:
                break
            # Показалась проверка Cloudflare — кликаем по виджету (и человеку есть куда нажать).
            await solve_turnstile(page, timeout=25)

        if not raw:
            title = await page.title()
            text = (await page.evaluate("() => document.body.innerText.slice(0, 160)")) or ''
            print(json.dumps({'ok': False, 'reason': 'число не найдено',
                              'title': title, 'text': re.sub(r'\s+', ' ', text)},
                             ensure_ascii=False))
            return 1

        digits = re.sub(r'[^\d.]', '', str(raw))
        print(json.dumps({'ok': True, 'balance': float(digits) if '.' in digits else int(digits),
                          'raw': raw}, ensure_ascii=False))
        return 0


if __name__ == '__main__':
    sys.exit(asyncio.run(main()))

File: bai/read-balance.js
// Читает остаток с chat.b.ai/usage браузером профиля шлюза.
//
// Зачем браузером: страница за Cloudflare, и любой HTTP-запрос со стороны (curl, серверный fetch,
// даже с живыми куками профиля) получает `403 Cf-Mitigated: challenge` вместо данных — проверено
// 15.09. Профиль шлюза логинится в chat.b.ai через Google, поэтому читаем в самом браузере.
//
// Печатает одну строку JSON: {"ok":true,"balance":297604,"raw":"297,604"} или причину отказа.
// Запуск: node bai/read-balance.js [--visible]   (--visible — если Cloudflare не пропустит headless)
const path = require('path');
const { chromium } = require('playwright');

const PROFILE = path.join(__dirname, 'profiles');
const URL = 'https://chat.b.ai/usage';

(async () => {
    const visible = process.argv.includes('--visible');
    const realChrome = process.argv.includes('--chrome');   // настоящий Chrome: Cloudflare к нему добрее
    const ctx = await chromium.launchPersistentContext(PROFILE, {
        ...(realChrome ? { channel: 'chrome' } : {}),
        headless: !visible,
        viewport: visible ? null : { width: 1280, height: 900 },
        args: ['--disable-blink-features=AutomationControlled'],
    });
    try {
        const page = ctx.pages()[0] || await ctx.newPage();
        await page.goto(URL, { waitUntil: 'domcontentloaded', timeout: 60000 });

        // 🪤 Классы у элементов сгенерированные (`acss-1h7d5nm`) и меняются от сборки к сборке,
        // поэтому цепляемся не за класс, а за СОСЕДА элемента с текстом ровно «Balance».
        // Ждём долго: если Cloudflare показал проверку, человеку нужно время её пройти.
        const handle = await page.waitForFunction(() => {
            const label = [...document.querySelectorAll('div')]
                .find(n => n.children.length === 0 && n.textContent.trim() === 'Balance');
            const txt = label && label.nextElementSibling && label.nextElementSibling.textContent.trim();
            return txt && /\d/.test(txt) ? txt : null;
        }, { timeout: Number(process.env.BALANCE_WAIT_MS || 150000) }).catch(() => null);

        const raw = handle ? await handle.jsonValue() : null;
        if (!raw) {
            // Диагностика в отказе: «челлендж Cloudflare» и «не залогинен» лечатся разным.
            const title = await page.title().catch(() => '');
            const text = await page.evaluate(() => document.body.innerText.slice(0, 160).replace(/\s+/g, ' ')).catch(() => '');
            console.log(JSON.stringify({ ok: false, reason: 'число не найдено', url: page.url(), title, text }));
            process.exit(1);
        }
        console.log(JSON.stringify({ ok: true, balance: Number(String(raw).replace(/[^\d.]/g, '')), raw }));
    } finally {
        await ctx.close();
    }
})().catch(e => {
    console.log(JSON.stringify({ ok: false, reason: String(e.message).slice(0, 140) }));
    process.exit(1);
});
